feat(rds): aurora - iam role to import and export data from s3

PikeDev opened the pull request:
## Commit Message
feat(rds): aurora - iam role to import and export data from s3 (#6611)

closes #6610 
## End Commit Message
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

<!-- 
Please read the contribution guidelines and follow the pull-request checklist:
https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md
 -->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 36773d109c947270a3bbba6c733770847bc66b32
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=01ebeade-1ed3-4648-88d7-353a9831502d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7a2721626f4a4a96b62f6aed16704c0ab566c6bb
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=046a2b4c-d5e5-431c-ad82-85be917ace0c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d85ac094aec17ba61a42df16907f56484499caea
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f3cfd13d-c069-4655-b768-9e60349ce963%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 48a71c58c02078cf3951b6abf8944e04efc0c2c3
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=40a63a4d-2cce-4cb0-bca5-644a785622c7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 490a97084bc014554f05ce92019f07b5b3514c05
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1f083560-3d74-44d4-ab85-e0671b626fab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

PikeDev commented:
Sorry for the many commits in a row. Should be good for review now.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 859bb433cffe6471bf8453b911a676ac1471e7d3
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e1eaa7a0-f25d-482f-a515-494ffeb3a79c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d8b638d86c784b4652953cfcb2255c6bcfe54843
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=10182868-108f-4ebb-b6ae-91c4c15e18cc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 85ed69a05fe27562f51fe42fe96ca7ae9a3444af
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0ae8908e-40c9-426c-bc62-58b16c1d288a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 16471093baa4943857c7fc0ee5282c4c20dac258
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5a8b5380-acd9-4bce-bd19-ca6d1ec196d7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

PikeDev commented:
Hi @nija-at. I agree with you and applied the changes. Note that I left out the optional feature name because no matter what I specify (s3Import, s3import etc.), CloudFormation tells me the feature doesn't exist. Without the optional parameter I do get the required functionality so it should be fine.

Although the docs only specify s3Import you can actually specify two different roles for "SELECT INTO S3" (aurora_select_into_s3_role) and "LOAD DATA FROM S3" (aurora_load_from_s3_role). Do you think we should support this by adding another property s3ExportRole?

On another note, we could also set the role in the cluster's parameter group (as aws_default_s3_role) when the s3ImportRole property is set. Do you think that's a good thing to do or should the user just handle the change to the parameter group separately?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ff2629870ca9abf793d362f172bcd448a32174b2
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=685a7381-42b6-49fa-94a3-7c819f483b1f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e9becaac5888eedc194545d6ab7b15b7717a1b00
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3d9a3e9d-ee8b-4125-9e74-9d2d17605f31%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented:
> Although the docs only specify s3Import you can actually specify two different roles for "SELECT INTO S3" (aurora_select_into_s3_role) and "LOAD DATA FROM S3" (aurora_load_from_s3_role). Do you think we should support this by adding another property s3ExportRole?

Can you provide links to these docs? And any docs that describe the s3 import functionality.

PikeDev commented:
[Loading Data into an Amazon Aurora MySQL DB Cluster from Text Files in an Amazon S3 Bucket](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Integrating.LoadFromS3.html)
[Saving Data from an Amazon Aurora MySQL DB Cluster into Text Files in an Amazon S3 Bucket](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Integrating.SaveIntoS3.html)
[DBClusterRole docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-rds-dbcluster-dbclusterrole.html)

nija-at commented:
> Although the docs only specify s3Import you can actually specify two different roles for "SELECT INTO S3" (aurora_select_into_s3_role) and "LOAD DATA FROM S3" (aurora_load_from_s3_role). Do you think we should support this by adding another property s3ExportRole?

Quoting from the [documentation here](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Integrating.LoadFromS3.html) - "Set either the aurora_load_from_s3_role or aws_default_s3_role DB cluster parameter to the Amazon Resource Name (ARN) of the new IAM role."

It seems that is a property that is set as a DB cluster parameter, why do you think this can be set under `AssociatedRoles`?

nija-at commented on packages/@aws-cdk/aws-rds/README.md:
```diff
@@ -43,6 +43,17 @@ By default, the master password will be generated and stored in AWS Secrets Mana
 
+### Enabling S3 integration to a cluster
+
+To enable S3 integration on a clustered database, set an IAM role with the required S3 policies to the s3ImportRole property. Remember to also set this role as the aws_default_s3_role, aurora_load_from_s3_role and/or aurora_select_into_s3_role in the cluster's parameter group.
```
Use line width of 120.

Can you link to documentations where a user can learn more about what this s3 integration is and how to set it up? Usually, written as "Read more about S3 integration [here](link)".

Can you also link to details on why the - `aws_default_s3_role`, `aurora_load_from_s3_role` and `aurora_select_into_s3_role` need to be set? 
The latter two suggests that this is an Aurora only feature. Is that the case? Is this supported for all database types?

PikeDev replied:
It is indeed Aurora only. I've now added that to the readme. Also included links to the docs for both import and export.

PikeDev commented:
> 
> Quoting from the [documentation here](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Integrating.LoadFromS3.html) - "Set either the `aurora_load_from_s3_role` or `aws_default_s3_role` DB cluster parameter to the Amazon Resource Name (ARN) of the new IAM role."
> 
> It seems that is a property that is set as a DB cluster parameter, why do you think this can be set under `AssociatedRoles`?

There are two steps involved to enable S3 import (load from S3):
1. Add Role ARN to cluster parameter group
2. Associate role from step 1 with Cluster (e.g. through AssociatedRoles when using CFN)

Same thing goes for S3 export (select into S3), which can be a different role. For example one role with only S3 read permissions and one with S3 write permissions. So now you have specified 2 different roles in the cluster parameter group and you have to associate both with the Cluster to enable them. With only one role property this is not possible. Hope that makes sense.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 33b9346e86e34265ddeccc1439fbf2c2acf38d88
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=11f9288b-c2ab-4377-b8b2-79398505aaad%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented:
@PikeDev -

Thanks for giving a brief explanation on this. I'm not intimately familiar with all of the RDS Aurora features. 

If that's the case, the CDK should do both of this for the user, i.e., add the role ARN to the cluster parameter group, as well as, associate the role with the cluster.

Further, I suggest we create two properties - one that takes the role (as you've already done) and the other that takes a list of `s3.Bucket` objects. If the latter is specified, the CDK would create this IAM role and attach it to the appropriate properties. However, the user can always add a role that they hand create (if they wish to) using the former property.

What do you think?

There might be a bit of a challenge implementing this, since the property will need to be configured on `DatabaseCluster` construct and get added to the `ParameterGroup` construct. I'm here to help out if you get stuck or into any complications.

PikeDev commented:
@nija-at I definitely agree. I have added the buckets property which creates the role and parameter group. I have also added separate sets of properties for import and export.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f9380be8453a85b486ecfd35fa3c4e5cfbfc2cb9
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c0b9fc7b-010a-451c-873d-19db6cd54898%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d54590ab83da00df1e7dcd906a52b7d5fbe0eefe
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d9df365a-ffff-47c6-b79f-643951953bba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 88303caa246e7895df390efe649103672b6b9709
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ea3b252d-c2ab-432e-95f0-9f9c582391b8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -141,6 +142,34 @@ export interface DatabaseClusterProps {
+   *
+   * @default - A role is created for you if the s3ImportBuckets property is set
+   */
+  readonly s3ImportRole?: IRole;
```
All these properties are only applicable to Aurora, IIUC. 

It would be worth adding a statement to this effect here, and validations in the implementation.

PikeDev replied:
Done.  BTW I do agree with #6713.  Will make things quite cleaner.

PikeDev commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +337,80 @@ export class DatabaseCluster extends DatabaseClusterBase {
+    let s3ImportRole = props.s3ImportRole;
+    if (props.s3ImportBuckets && props.s3ImportBuckets.length > 0) {
+      if (props.s3ImportRole) {
+        throw new Error(`Property s3ImportRole should not be specified when specifying s3ImportBuckets. A role with permissions for these buckets will be created automatically`);
```
Shortened the error altogether

PikeDev commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +337,80 @@ export class DatabaseCluster extends DatabaseClusterBase {
+                  resources: getBucketArns(props.s3ImportBuckets)
+              })
+            ]
+          }),
```
Good catch - added

PikeDev commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +337,80 @@ export class DatabaseCluster extends DatabaseClusterBase {
+          parameters: {
+            aurora_load_from_s3_role: s3ImportRole ? s3ImportRole.roleArn : '',
+            aurora_select_into_s3_role: s3ExportRole ? s3ExportRole.roleArn : ''
+          }
```
Changed it to only add parameters that are actually set

PikeDev commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +337,80 @@ export class DatabaseCluster extends DatabaseClusterBase {
+      }
+
+      if (clusterParameterGroup) {
+        this.node.addWarning('Can not set S3 role(s) to your custom cluster parameter group. Make sure to set aurora_load_from_s3_role and/or aurora_select_into_s3_role parameters manually.');
```
Done. The whole thing has improved with this change. Specifying a handmade role will also modify the parameter group, be it a custom group or automatically generated.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ed4671f84a22fdcfb1e27696cfa2fec18795bb78
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4efc3c7b-3b95-4852-8eb9-45629600e205%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ed6039d19caed9fe59ffa27ef4025724c797d9ec
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0ebe3df9-fd9b-4814-8615-6ff41baf4f36%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 197ec3665d9697a1f515ad46a78805104035b123
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1fd0e61d-7605-4e4c-9729-81c6f569359c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 39225410040be156d0de43f44efa11ca0ed23562
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4d681f37-65aa-4f84-ac4a-f8ee62a511c1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

PikeDev commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +371,86 @@ export class DatabaseCluster extends DatabaseClusterBase {
+      }
+      if (s3ExportRole) {
+        clusterAssociatedRoles.push({ roleArn: s3ExportRole.roleArn });
+      }
```
Yup, tested it and it works great. I have yet to find a FeatureName that gets accepted by CloudFormation. I'm guessing it is just informational in the Console.

PikeDev commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +371,86 @@ export class DatabaseCluster extends DatabaseClusterBase {
+      }
+      if (s3ExportRole) {
+        clusterAssociatedRoles.push({ roleArn: s3ExportRole.roleArn });
+      }
```
Unfortunately not. I had them together in a previous commit, but then I realised the parameter group part is only relevant for MySQL. PostgreSQL only requires the role to be associated.

nija-at replied:
This is weird behaviour from RDS, but ok.

Can we add some comments in the code to mention this (and link to any docs, if any)?

I haven't looked into your tests yet and, if not already, let's capture them as unit or integration tests.

PikeDev replied:
Added a comment and an additional test for PostgreSQL.

PikeDev commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +371,86 @@ export class DatabaseCluster extends DatabaseClusterBase {
+      });
+    }
+
+    let clusterParameterGroup = props.parameterGroup;
```
I'm not sure where that would be. It is used in the block directly below.

nija-at replied:
I was simply suggesting to move it to line 438. Minor readability, not a major comment.

PikeDev replied:
Ah ok, gotcha. The field is also used further down so prefer to leave as is.

PikeDev commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +371,86 @@ export class DatabaseCluster extends DatabaseClusterBase {
+        clusterAssociatedRoles.push({ roleArn: s3ExportRole.roleArn });
+      }
+
+      if (props.engine === DatabaseClusterEngine.AURORA || props.engine === DatabaseClusterEngine.AURORA_MYSQL) {
```
Yes, the rest of the code works for all `DatabaseClusterEngine` types.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e3c941ce3c9f5f8b249922b6d42162c1a416bc8e
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f4f23263-6c1a-4f88-816d-ca68bcc8c3ba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: fad71e190b17e9c2485bb5c70faad3b6db237dfd
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d6235385-6e08-4914-b217-35fff4b9559b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 49491406d8fe23d1134fc693a85f649eef6ddc19
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bef15787-c947-4156-b45a-9e2d4a6ae668%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +371,86 @@ export class DatabaseCluster extends DatabaseClusterBase {
+        if (!clusterParameterGroup) {
+          clusterParameterGroup = new ClusterParameterGroup(this, "ClusterParameterGroup", {
+            family: getClusterParameterGroupFamily(props.engine, props.engineVersion),
+            parameters: {}
```
Now that `parameters` is optional, it can be dropped from here, yes?

nija-at commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -459,3 +604,38 @@ export class DatabaseCluster extends DatabaseClusterBase {
+    return "aurora-postgresql11";
+  }
+  throw new Error(`Unknown database engine: ${engine.name}`);
+}
```
This doesn't feel like the right place for this. It should go into `DatabaseClusterEngine` and `DatabaseInstanceEngine`.

nija-at commented on packages/@aws-cdk/aws-rds/lib/parameter-group.ts:
```diff
@@ -21,7 +29,7 @@ abstract class ParameterGroupBase extends Resource implements IParameterGroup {
    */
   public static fromParameterGroupName(scope: Construct, id: string, parameterGroupName: string): IParameterGroup {
-    class Import extends Resource implements IParameterGroup {
+    class Import extends ParameterGroupBase implements IParameterGroup {
```
Why modify this? Imported parameter groups are ones that are defined elsewhere but are imported into the CDK as read-only. It's not possible to modify an imported parameter group.

PikeDev replied:
I didn't want to define an empty `addParameter` method there and since that's now part of `IParameterGroup` it's required. Do you have any idea to solve this neatly or should I just opt for that empty method?

nija-at replied:
Ah, I see what happened. 

The better to way to fix this is -
1. drop the `addParameter` method on the `IParameterGroup` interface.
2. change the code at line 439 of `cluster.ts` to 
   ```ts
    if (!clusterParameterGroup) {
      clusterParameterGroup = new ClusterParameterGroup(this, "ClusterParameterGroup", {
        family: props.engine.getClusterParameterGroupFamily(props.engineVersion)
      });
    } else if (clusterParameterGroup instanceof ClusterParameterGroup) { // Skip modification on an imported parameter group
      if (s3ImportRole) {
        clusterParameterGroup.addParameter('aurora_load_from_s3_role', s3ImportRole.roleArn);
      }
      if (s3ExportRole) {
        clusterParameterGroup.addParameter('aurora_select_into_s3_role', s3ExportRole.roleArn);
      }
    }
   ```

Does that make sense?

nija-at commented on packages/@aws-cdk/aws-rds/test/integ.cluster.ts:
```diff
@@ -1,23 +1,19 @@
+
+const importBucket = new s3.Bucket(stack, 'ImportBucket');
+const exportBucket = new s3.Bucket(stack, 'ExportBucket');
+
```
Avoid modifying existing integration tests. Add new ones instead.

Optionally, if there is a simple way to verify that s3 import and export works, after deploying the stack, document that too. See [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#integration-tests) for some examples.

PikeDev replied:
Done.

The verification steps are a bit too comprehensive I think. First need to get access through a tunnel or similar, then create a table, insert data, export into bucket, import it back into database etc.

nija-at commented on packages/@aws-cdk/aws-rds/test/test.cluster.ts:
```diff
@@ -516,6 +517,425 @@ export = {
+          ]
+        }
+      }]
+    }, ResourcePart.Properties));
```
Minor: `ResourcePart.Properties` is the default, so you can drop it.

nija-at commented on packages/@aws-cdk/aws-rds/test/test.cluster.ts:
```diff
@@ -516,6 +517,425 @@ export = {
+
+    test.done();
+  },
+  'create a cluster with s3 export role'(test: Test) {
```
Also assert that the IAM role is created with the expected permissions.

nija-at commented on packages/@aws-cdk/aws-rds/test/test.cluster.ts:
```diff
@@ -516,6 +517,425 @@ export = {
+
+    test.done();
+  },
+  'create a cluster with s3 import buckets'(test: Test) {
```
Also assert that the IAM role is created with the expected permissions.

nija-at commented on packages/@aws-cdk/aws-rds/test/test.cluster.ts:
```diff
@@ -516,6 +517,425 @@ export = {
+
+    test.done();
+  },
+  'create a cluster with s3 export buckets'(test: Test) {
```
Minor: new line between tests.

nija-at commented:
I want to add - thanks for all incorporating all the comments. There's a couple of more places that need adustments, per my comments.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 07dc4ee922aefe669589424de1a61233508ebc98
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=786387f6-9909-45b9-a23c-fa6b165a3cd1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 6210a1cadb9060a2f0c0012feb9fc7f4d78bbf85
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a9d1b491-98a9-43c7-821f-641cc41088e9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-rds/lib/parameter-group.ts:
```diff
@@ -51,8 +71,10 @@ export interface ParameterGroupProps {
   /**
    * The parameters in this parameter group
+   *
+   * @default None
```
```suggestion
   * @default - None
```

nija-at commented on packages/@aws-cdk/aws-rds/lib/props.ts:
```diff
@@ -36,6 +36,29 @@ export class DatabaseClusterEngine {
+      return "aurora-postgresql11";
+    }
+    throw new Error(`Unknown database engine: ${this.name}`);
+  }
```
Sorry to keep pressing on this, but the implementation needs to be overhauled.

The constructor of `DatabaseClusterEngine` needs to be modified to take in the list of parameter group families which is then stored as an instance property. This method itself then uses the optional `engineVersion` parameter to determine which of the parameter group family to be returned.

Remember to add a unit test.

PikeDev replied:
Don't worry, I understand. I have changed it as per your suggestion and I'd say it is indeed better, but still not great. It's a necessary evil I think. Best thing would be if the CFN team just makes the Family property optional to let RDS figure out the proper family. Perhaps in the future and then we can clean this up.

nija-at commented on packages/@aws-cdk/aws-rds/lib/props.ts:
```diff
@@ -36,6 +36,29 @@ export class DatabaseClusterEngine {
+  /**
+   * Get this engine's default parameter group family for given version
+   */
+  public getClusterParameterGroupFamily(engineVersion?: string): string {
```
1. Rename this to `defaultParameterGroupFamily`

2. Is there a link that you can add to the documentation which describes what the default parameter group family is? (I'm unfamiliar with the concept and I'd like to myself learn what a parameter group family is)

3. Is this *the* parameter group family, or only the default?

PikeDev replied:
I suppose it's  _the_ parameter group, at least at this point in time there seems to be only one family per engine version. I have renamed it as such.

jogold commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +371,86 @@ export class DatabaseCluster extends DatabaseClusterBase {
+            ]
+          }),
+        }
+      });
```
Suggestion: use `grantRead()`?

```suggestion
      s3ImportRole = new Role(this, "S3ImportRole", {
        assumedBy: new ServicePrincipal("rds.amazonaws.com")
      });
      for (const bucket of props.s3ImportBuckets) {
        bucket.grantRead(s3ImportRole);
      }
```

The logic in `bucketAndObjectArns` is already implemented in `grantRead()`

PikeDev replied:
Thanks, that does clean things up!

jogold commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +371,86 @@ export class DatabaseCluster extends DatabaseClusterBase {
+            ]
+          }),
+        }
+      });
```
```suggestion
      s3ExportRole = new Role(this, "S3ExportRole", {
        assumedBy: new ServicePrincipal("rds.amazonaws.com"),
      });
      for (const bucket of props.s3ExportBuckets) {
        bucket.grantReadWrite(s3ExportRole);
      }
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3799c8ce9902eab794ce1123d7a9ed7c769f6079
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bdce74c2-33cd-465f-92bb-9d2f3676619e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 06da7c648d6c6d9f39661315aefb2f0750ad9e14
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c48f689b-adc6-4a79-af90-61a8205069b7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-rds/lib/cluster.ts:
```diff
@@ -308,6 +371,63 @@ export class DatabaseCluster extends DatabaseClusterBase {
+          });
+        }
+
+        if (clusterParameterGroup instanceof ClusterParameterGroup) {
```
```suggestion
        if (clusterParameterGroup instanceof ClusterParameterGroup) { // ignore imported ClusterParameterGroup
```

PikeDev commented on packages/@aws-cdk/aws-rds/lib/props.ts:
```diff
@@ -30,11 +40,29 @@ export class DatabaseClusterEngine {
+    } else if (this.parameterGroupFamilies.length > 0) {
+      return this.parameterGroupFamilies[this.parameterGroupFamilies.length - 1][1];
+    }
+    throw new Error(`Could not determine parameter group family for database engine version: ${engineVersion}`);
```
This leads to trouble elsewhere because `family` is a required property of parameter group, even in CFN generated code.

PikeDev commented on packages/@aws-cdk/aws-rds/lib/props.ts:
```diff
@@ -30,11 +40,29 @@ export class DatabaseClusterEngine {
    */
   public readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
 
+  private readonly parameterGroupFamilies: Array<[string, string]>;
```
Had to export it due to:

> Cannot use private type ParameterGroupFamily in exported declarations

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 251fc81fef649ba6cbb084e083b5ce49a74911fd
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=befb3dad-b258-455a-8da2-4d212f888ecf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented:
I've gone ahead and pushed a commit to fix up [based on this comment](https://github.com/aws/aws-cdk/pull/6611#discussion_r397167074) and updated the behaviour of `parameterGroupFamily()`. 

Hope you don't mind. 😊

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d23362ec4ce774725e380c125c7158b7172663a3
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b82b1b79-0721-42f4-8395-f1391d223be9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

PikeDev commented:
Good stuff!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a33ee09cbf6bad001eca1095f51b4c4e8ce5ac85
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=44e0d049-ff3e-430b-be03-8734a7dcc858%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
